heartbeat/scheduler/timerqueue/queue.go

```diff
@@ -0,0 +1,163 @@
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package timerqueue
+
+import (
+	"container/heap"
+	"context"
+	"time"
+)
+
+// TimerTask represents a task run by the TimerQueue.
+type TimerTask struct {
+	fn    TimerTaskFn
+	runAt time.Time
+}
+
+// NewTimerTask creates a new TimerTask struct.
+func NewTimerTask(runAt time.Time, fn TimerTaskFn) *TimerTask {
+	return &TimerTask{runAt: runAt, fn: fn}
+}
+
+// TimerTaskFn is the function invoked by a TimerTask.
+type TimerTaskFn func(now time.Time)
+
+// TimerQueue represents a priority queue of timers.
+type TimerQueue struct {
+	th            *timerHeap
+	ctx           context.Context
+	nextRunAt     *time.Time
+	popRunnableCh chan chan []*TimerTask
+	pushCh        chan *TimerTask
+	timer         *time.Timer
+}
+
+// NewTimerQueue creates a new instance.
+func NewTimerQueue(ctx context.Context) *TimerQueue {
+	tq := &TimerQueue{
+		th:            &timerHeap{},
+		ctx:           ctx,
+		popRunnableCh: make(chan chan []*TimerTask),
+		pushCh:        make(chan *TimerTask),
+		timer:         time.NewTimer(0),
+	}
+	heap.Init(tq.th)
+
+	return tq
+}
+
+// Push adds a task to the queue
+func (tq *TimerQueue) Push(tt *TimerTask) {
+	// Block until push succeeds or shutdown
+	select {
+	case tq.pushCh <- tt:
+	case <-tq.ctx.Done():
+	}
+}
+
+// PopRunnable pops as many runnable tasks from the queue as possible
+func (tq *TimerQueue) PopRunnable() (res []*TimerTask) {
+	popDone := make(chan []*TimerTask)
+	tq.popRunnableCh <- popDone
+
+	select {
+	case res = <-popDone:
+		return res
+	case <-tq.ctx.Done():
+		return
+	}
+}
+
+// RunRunnableTasks runs all tasks that are currently runnable. Tasks are run in serial blocking manner in the
+// current go-routine
+func (tq *TimerQueue) RunRunnableTasks() {
+	runnable := tq.PopRunnable()
+	now := time.Now()
+	for _, tt := range runnable {
+		tt.fn(now)
+	}
+}
+
+// Start runs a goroutine within the given context that processes items in the queue, spawning a new goroutine
+// for each.
+func (tq *TimerQueue) Start() {
+	go func() {
+		for {
+			// flag controlling whether we should update tq.nextRunAt after the select block
+			updateNextRunAt := false
+
+			select {
+			case <-tq.ctx.Done():
+				// Stop the timerqueue
+				return
+			case <-tq.timer.C:
+				tq.RunRunnableTasks()
+				updateNextRunAt = true
+			case retCh := <-tq.popRunnableCh:
+				res := tq.popRunnableUnsafe()
+				updateNextRunAt = true
+				retCh <- res
+			case tt := <-tq.pushCh:
+				tq.pushUnsafe(tt)
+			}
+
+			if updateNextRunAt {
+				tq.updateTimer()
+			}
+		}
+	}()
+}
+
+func (tq *TimerQueue) updateTimer() {
+	if tq.th.Len() == 0 {
+		tq.nextRunAt = nil
+		tq.timer.Stop()
```

Comment:
Stopping timers can be subject to races, especially if a timer is reused via `Reset`. In case some element has been removed via pop and we are at 0 now, we still might have a timer that might have fired. Common pattern (according to go docs):

```
if !tq.timer.Stop() {
  <-tc.timer.C  // consume signal if timer did already fire
}
```

By not consuming from the channel we might reschedule a timer that is already marked as triggered and therefore fires immediately in the following select.

Reply:
I don't think we can have races here since we do all scheduling and receiving in a single thread.

Reply:
You still rely on the timer package for waiting. `tc.timer` has type `*time.Timer`. The Timer type is tricky to use. Once scheduled via `Reset` the timer will not run in your thread/go-routine. It is the timer sub-system that is going to put a signal into the channel. Which means that `Stop` and `Reset` are always subject to races with the actual timer subsystem. According to `time.Timer` docs one must call Reset only after calling Stop, and on Stop one must be prepared to consumer a pending signal from C.

Reply:
Thanks for the clarification, will fix!

Reply:
I've since fixed this, the semantics here are really tricky (looking at the history of go timers it looks like they've been the topic of a lot of confusion and debate).